### DISCUSSION TITLE
How to install poetry inside a docker container ?

### DISCUSSION THREAD

**Discussion by aahnik** [python-poetry/poetry#3933] [Q&A] [answered]
I read this https://github.com/python-poetry/poetry/discussions/1879#discussioncomment-216865 , but it seems too complex, for a beginner like me.

I tried, running 

```shell
$ curl -sSL https://raw.githubusercontent.com/python-poetry/poetry/master/get-poetry.py | python -
```
and this works perfectly in my machine (ubuntu) as well and docker containers running linux.

But the thing is that, the above command requires to run `source $HOME/.poetry/env` to access the `poetry` command.

For example:

I create a Dockerfile:

```Dockerfile
FROM python:3.9

WORKDIR /app

RUN apt-get update && apt-get upgrade -y

RUN curl -sSL https://raw.githubusercontent.com/python-poetry/poetry/master/get-poetry.py | python -

RUN bash -c "source $HOME/.poetry/env"

# RUN poetry --help # will be successful

# RUN poetry install # will succeed

CMD ["bash"]

```
Now I build the image:

```shell
$ docker build --tag poetry-docker .
```

Now, I run it in interactive mode:

```shell
$ docker run -it poetry-docker
root@bb4c99c703e4:/app# poetry --version
bash: poetry: command not found
root@bb4c99c703e4:/app# source $HOME/.poetry/env
root@bb4c99c703e4:/app# poetry --version
Poetry version 1.1.5
```
If I stop the container, and start again, I again need to run the source command, to be able to use `poetry` command.

----------

My questions:

- is the Dockerfile I made, safe to use for real applications?
- why do I need to `source $HOME/.poetry/env` every time? **How to add `poetry` to the path, once and for all?**

basically, i want to publish my poetry-docker image to Dockerhub, and want to use it in my future projects like this:

```Dockerfile
FROM aahnik/poetry-docker

RUN poetry --help # should succeed

...
```

-----
Update: Here https://python-poetry.org/docs/ get-poetry.py is used. But in README, it says deprecated?

is my problem related to that ?

**Comment by aahnik** [comment]
I changed my Dockerfile (see original above) to this, and it solves my issue of " How to add poetry to the path, once and for all?". But I am curious to know if my Dockerfile is all correct and safe for production.

```diff
-RUN curl -sSL https://raw.githubusercontent.com/python-poetry/poetry/master/get-poetry.py | python -
+RUN curl -sSL https://raw.githubusercontent.com/python-poetry/poetry/master/install-poetry.py | python -
 
-RUN bash -c "source $HOME/.poetry/env"
+ENV PATH /root/.local/bin:$PATH
```

**Reply by abn** [reply]
For completeness, you can also the export is done via `/etc/profile.d/local-bin.sh` or something. Not necessarily required in most cases.

**Reply by abn** [reply]
However, the `/root/.local/bin` dir should alredy be present in the shell path.

**Comment by abn** [comment, marked as answer]
> * is the Dockerfile I made, safe to use for real applications?

There is nothing obvious that is wrong here. Maybe cleanup after your `apt-get upgrade`.

> * why do I need to `source $HOME/.poetry/env` every time? **How to add `poetry` to the path, once and for all?**

This is because `source <file>` only applies to the current shell, and each `RUN` executes in a new shell. To make it permanent, you can add the command to `/etc/profile.d/poetry.sh` and also add bash completions if you so desired to `/etc/bash_completion.d/poetry`. 

> Update: Here https://python-poetry.org/docs/ get-poetry.py is used. But in README, it says deprecated?

The live docs are related to the current stable version. The documentation on `master` refers to the latest in-development doc. With `1.2.0`, the `install-poetry.py` script will become the recommendation. Note that if you use `install-poetry` today, it will install `1.1.6`, this version is not aware of the new installation mechanism and hence if you try `poetry self update` it will fail (this is the only draw back).

In a container you can also use `python -m pip install poetry`. The side-effect here is that poetry gets installed in the system-stite, re-using any existing dependencies. In this case, make sure that you do not use the system-site for project as well (ie. use a virtual envrionment for your project - always a good practice), as this may cause other issues.

**Reply by aahnik** [reply]
thanks a lot. now i understand that `pip install poetry` is completely safe, if I don't have other packages at system level.

`pip install poetry` is good to go for github actions and docker containers, if I run my application inside virtual environment, (which poetry does by default.)

thank you for such detailed explanation, 🙏

**Reply by rashid200gb** [reply]
Bundle of thanks for valuable information.
